scene/animation/animation_node_state_machine.cpp

```diff
@@ -751,58 +749,50 @@ AnimationNode::NodeTimeInfo AnimationNodeStateMachinePlayback::_process(const St
-		Vector<StringName> new_path;
-		bool can_travel = _make_travel_path(tree, p_state_machine, travel_path.size() <= 1 ? p_state_machine->is_allow_transition_to_self() : false, new_path, p_test_only);
-		if (travel_path.size()) {
+
+		// If we are already planning to teleport, we can skip the expensive A* attempt.
+		if (!teleport_request) {
+			// Fix path.
+			// Process children.
+			Vector<StringName> new_path;
+			bool can_travel = _make_travel_path(tree, p_state_machine, travel_path.size() <= 1 ? p_state_machine->is_allow_transition_to_self() : false, new_path, p_test_only);
+			if (travel_path.size()) {
+				if (can_travel) {
+					can_travel = _travel_children(tree, p_state_machine, travel_target, p_state_machine->is_allow_transition_to_self(), travel_path[0] == current, p_test_only);
+				} else {
+					_start_children(tree, p_state_machine, travel_target, p_test_only);
+				}
+			}
+
+			// Process to travel.
 			if (can_travel) {
-				can_travel = _travel_children(tree, p_state_machine, travel_target, p_state_machine->is_allow_transition_to_self(), travel_path[0] == current, p_test_only);
+				path = new_path;
 			} else {
-				_start_children(tree, p_state_machine, travel_target, p_test_only);
+				// Can't travel, then teleport.
+				if (p_state_machine->states.has(temp_travel_request)) {
+					teleport_request = true;
+				} else {
+					ERR_FAIL_V_MSG(AnimationNode::NodeTimeInfo(), "No such node: '" + temp_travel_request + "'");
+				}
 			}
 		}
 
-		// Process to travel.
-		if (can_travel) {
-			path = new_path;
-		} else {
-			// Can't travel, then teleport.
-			if (p_state_machine->states.has(temp_travel_request)) {
-				path.clear();
-				if (current != temp_travel_request || reset_request_on_teleport) {
-					_set_current(p_state_machine, temp_travel_request);
-					reset_request = reset_request_on_teleport;
-					teleport_request = true;
-				}
-			} else {
-				ERR_FAIL_V_MSG(AnimationNode::NodeTimeInfo(), "No such node: '" + temp_travel_request + "'");
+		if (teleport_request) {
+			path.clear();
+			if (p_state_machine->is_allow_transition_to_self() || current != temp_travel_request) {
+				path.push_back(temp_travel_request);
 			}
+			_transition_to_next_recursive(tree, p_state_machine, p_delta, p_test_only);
+			teleport_request = false;
 		}
 	}
 
 	AnimationMixer::PlaybackInfo pi = p_playback_info;
 
-	if (teleport_request) {
-		teleport_request = false;
-		// Clear fadeing on teleport.
-		fading_from = StringName();
-		fadeing_from_nti = AnimationNode::NodeTimeInfo();
-		fading_pos = 0;
-		// Init current length.
-		pi.time = 0;
-		pi.seeked = true;
-		pi.is_external_seeking = false;
-		pi.weight = 0;
-		current_nti = p_state_machine->blend_node(p_state_machine->states[current].node, current, pi, AnimationNode::FILTER_IGNORE, true, true);
-		// Don't process first node if not necessary, instead process next node.
-		_transition_to_next_recursive(tree, p_state_machine, p_delta, p_test_only);
-	}
-
```

Comment:
However, I think it is still necessary to add some kind of detailed distinction enum or additional bool flag internally for the teleport_request inside c++. This is because transitions from the start state and transitions by restart should not be treated as teleports.

Since playback by start() (i.e., immediate teleport) discards the currently playing transition here, this line would be required with some flags even if the flag `teleport_request` will be separated in detail.